### PR TITLE
Attach to framework tests

### PR DESCRIPTION
**Pull request by ofirgo**
## Pull Request Description:

Unit tests for the Pytorch and Keras attach TPC to framework modules.
Also removing old TPC error tests that where already commented out.

## Checklist before requesting a review:
- [ ] I set the appropriate labels on the pull request.
- [ ] I have added/updated the release note draft (if necessary).
- [ ] I have updated the documentation to reflect my changes (if necessary).
- [ ] All function and files are well documented. 
- [ ] All function and classes have type hints.
- [ ] There is a licenses in all file.
- [ ] The function and variable names are informative. 
- [ ] I have checked for code duplications.
- [ ] I have added new unittest (if necessary).

### CODE REVIEW

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+        filter_op = LayerFilterParams(op, kernel_size=1)
+
+        fw_quant_capabilities = self.attach2fw.attach(tpc,
+                                                      custom_opset2layer={opset_name: CustomOpsetLayers(
+                                                          operators=[filter_op])})
+
+        opset_layers = fw_quant_capabilities.op_sets_to_layers.get_layers_by_op(operator_set)
+        assert op not in opset_layers
+        assert filter_op in opset_layers
```
Comment (irenaby) [inline tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:243]: I think there should be a test with more than one op

File: tests_pytest/keras/target_platform_capabilities/targetplatform2framework/test_attach2keras.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    def test_attach2fw_attach_to_default_config(self):
+        super().test_attach2fw_attach_to_default_config()
+
+    def test_not_existing_opset_with_layers_to_attach(self):
+        super().test_attach2fw_attach_to_default_config()
+
+    def test_attach2pytorch_attach_with_custom_opset(self):
+        self._test_attach2fw_attach_with_custom_opset([Identity],
+                                                      LayerFilterParams(Conv2D, stride=2),
+                                                      KERAS_KERNEL)
```
Comment (irenaby) [inline tests_pytest/keras/target_platform_capabilities/targetplatform2framework/test_attach2keras.py:54]: Can it be any string?

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    attach2fw_class = None
+
+    def setup_method(self):
+        assert self.attach2fw_class is not None
+        self.attach2fw = self.attach2fw_class()
+
+
+    def test_attach2fw_init(self):
+        # verify built-in opset to operator mapping structure
+        assert all([opset in self.attach2fw._opset2layer for opset in schema.OperatorSetNames.get_values()])
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:53]: > This is confusing, k in dict checks if the dict has a key k, but you compare it to values, not keys. Is this true because OpSetNames is an str enum? Or do I miss something silly? Also you don't need the explicit list

The keys in _opset2layer are of type OperatorSetNames, as far as I understand it works because it is a str-enum yes.
I compare it to the values of the enum, I don't care for the values in the dict (which are lists of operators)

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+
+            all_mapped_ops = fw_quant_capabilities.layer2qco.copy()
+            all_mapped_ops.update(fw_quant_capabilities.filterlayer2qco)
+            if len(opset.op_list) == 0:
+                assert len(all_mapped_ops) == 0
+            else:
+                assert len(all_mapped_ops) == len(opset.op_list)
+
+                for qco in all_mapped_ops.values():
+                    assert len(qco.quantization_configurations) == 1
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:91]: qco.base_config should.
here I check the qco which contains both quantization_configurations and base_config.
not sure it is necessary but it can't harm.
I can change the assert in the line below to assert comparison to tested_op_cfg instead of asserting just the activation_n_bits

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+            weights_n_bits=8,
+            weights_per_channel_threshold=False,
+            enable_weights_quantization=False,
+            lut_values_bitwidth=None)
+
+        tested_attr_cfg = default_attr_config.clone_and_edit(weights_n_bits=42)
+
+        tested_op_cfg = default_op_cfg.clone_and_edit(default_weight_attr_config=default_attr_config,
+                                                      attr_weights_configs_mapping={KERNEL_ATTR: tested_attr_cfg,
+                                                                                    BIAS_ATTR: tested_attr_cfg})
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:108]: I don't really care about distinguishing between them, and it complicates the assertion (because, in order to get the attributes after the "attach" I'll need to use the attribute name which is fw-dependent).
So, I prefer to not get into this

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+            else:
+                assert len(all_mapped_ops) == len(opset.op_list)
+
+                for qco in all_mapped_ops.values():
+                    assert len(qco.quantization_configurations) == 1
+                    assert qco.base_config.default_weight_attr_config == default_attr_config
+
+                    for attr_name, fw_layer2attr_mapping in fw_linear_attr_names.items():
+                        assert isinstance(fw_layer2attr_mapping, DefaultDict)
+                        layer_attr_mapping = fw_layer2attr_mapping.get(opset.op_list[0])
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:142]: it's a DefaultDict I have to use get it is the only way to access the dict.
I'll change the next line

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+        operator_set = schema.OperatorsSet(name=opset_name)
+
+        tpc = schema.TargetPlatformCapabilities(default_qco=default_qc_options,
+                                                operator_set=tuple([operator_set]))
+
+        fw_quant_capabilities = self.attach2fw.attach(tpc)
+
+        assert isinstance(fw_quant_capabilities, FrameworkQuantizationCapabilities)
+        opset2layer = fw_quant_capabilities.op_sets_to_layers.get_layers_by_op(operator_set)
+        assert len(opset2layer) > 0
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:158]: not necessarily there can be more than 1 layer in an opset

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        opset_cfg = fw_quant_capabilities.layer2qco.get(opset2layer[0])
+        assert opset_cfg is not None
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:160]: I don't (will remove)

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+
+    def test_not_existing_opset_with_layers_to_attach(self):
+        opset_name = "NotExisting"
+        operator_set = schema.OperatorsSet(name=opset_name)
+
+        tpc = schema.TargetPlatformCapabilities(default_qco=schema.QuantizationConfigOptions(
+            quantization_configurations=(default_op_cfg,)),
+            operator_set=tuple([operator_set]))
+
+        with pytest.raises(Exception) as e_info:
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:171]: no type :(
and thanks

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+            _ = self.attach2fw.attach(tpc)
+        assert f'{opset_name} is defined in TargetPlatformCapabilities' in str(e_info)
+
+    def _test_attach2fw_attach_with_custom_opset(self, operators_list, filter_op, kernel_attr_name):
+        test_bit = 42
+        cfg = default_op_cfg.clone_and_edit(attr_weights_configs_mapping=
+                                            {KERNEL_ATTR: schema.AttributeQuantizationConfig(weights_n_bits=test_bit)})
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:178]: he can define whatever he wants as long as he passes a matching mapping in the `custom_opset2layer` (when calling attach).
So I'll change it in the test to not use KERNEL_ATTR just so we'll have this covered

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+        # Setting a layers mapping for the custom opset with a regular operator and a filter.
+        # We also test the option of passing an attributes mapping for the operator to set a specific attribute config.
+        custom_attr_name = 'CustomAttr'
+
+        assert len(operators_list) == 1
+        non_filter_op = operators_list[0]
+        fw_quant_capabilities = self.attach2fw.attach(
+            tpc,
+            custom_opset2layer={opset_name: CustomOpsetLayers(operators=[non_filter_op, filter_op],
+                                                              attr_mapping={KERNEL_ATTR: DefaultDict(
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:201]: I don't think it is necessary.
I didn't get the "Custom attr" part of your comment

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+                                                                  {filter_op: custom_attr_name},
+                                                                  default_value=kernel_attr_name)})
+                                })
+
+        assert isinstance(fw_quant_capabilities, FrameworkQuantizationCapabilities)
+        opset_to_layers = fw_quant_capabilities.op_sets_to_layers.op_sets_to_layers
+        assert len(opset_to_layers) == 1
+        assert opset_to_layers[0].name == opset_name
+        assert len(opset_to_layers[0].layers) == 2
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:210]: I don't understand. Which object should I compare opsets_to_layers  to?

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 13]
```diff
@@ -0,0 +1,243 @@
...
+
+        op_cfg = fw_quant_capabilities.layer2qco[non_filter_op].base_config
+        assert op_cfg.activation_n_bits == test_bit
+        assert kernel_attr_name in op_cfg.attr_weights_configs_mapping
+        assert op_cfg.attr_weights_configs_mapping[kernel_attr_name].weights_n_bits == test_bit
+
+        op_cfg = fw_quant_capabilities.filterlayer2qco[filter_op].base_config
+        assert op_cfg.activation_n_bits == test_bit
+        assert custom_attr_name in op_cfg.attr_weights_configs_mapping
+        assert op_cfg.attr_weights_configs_mapping[custom_attr_name].weights_n_bits == test_bit
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:220]: It really complicates the test to use a different config, I don't think it is important because I test them independently, no interaction between them is relevant for the test

File: tests_pytest/keras/target_platform_capabilities/targetplatform2framework/test_attach2keras.py (hunk 2 above)
Comment (ofirgo) [reply tests_pytest/keras/target_platform_capabilities/targetplatform2framework/test_attach2keras.py:54]: no it is operator and fw dependent

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (irenaby) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py]: ok, but is it intentional that you compare to enum values and not enums? If yes, please add a comment. It's very confusing that you compare to values and in the next line the type should be of enum (and not of values)

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (irenaby) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:108]: If you never look inside the object that might be fine, but there is any logic that accesses specific attributes, its better to distinguish. In general it's a bad habit to assign undistinguishable values to different things in tests since it might reduce the capability of the test to identify silly mistakes

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (irenaby) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:156]: In general or in this test?

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (irenaby) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py]: I meant why would a custom op have a known kernel attr?

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (irenaby) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:208]: I think you have the exact values that should be in the list. The point is to avoid loose checks if you can do an exact check, unless there is a good reason to. Again, less room for undiscovered bugs.

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+
+        assert isinstance(fw_quant_capabilities, FrameworkQuantizationCapabilities)
+        opset_to_layers = fw_quant_capabilities.op_sets_to_layers.op_sets_to_layers
+        assert len(opset_to_layers) == 1
+        assert opset_to_layers[0].name == opset_name
+        assert len(opset_to_layers[0].layers) == 2
+
+        op_cfg = fw_quant_capabilities.layer2qco[non_filter_op].base_config
+        assert op_cfg.activation_n_bits == test_bit
+        assert kernel_attr_name in op_cfg.attr_weights_configs_mapping
```
[the remark this answers is not in the document]
Comment (irenaby) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:212]: its a matter of preference I suppose, I prefer my tests less cluttered when the failure is clear

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py (hunk 1 above)
Comment (irenaby) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:241]: There should be some code that iterates over them, just to check it happens and there is no mixup. But if you think its not needed that's fine.

File: tests_pytest/base_test_classes/base_tpc_attach2fw_test.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ofirgo) [reply tests_pytest/base_test_classes/base_tpc_attach2fw_test.py:208]: compare to [non_filter_op, filter_op]

**Review by irenaby — APPROVED** [review]

**Review by lapid92 — APPROVED** [review]

**Review by lior-dikstein — APPROVED** [review]

**Review by Idan-BenAmi — APPROVED** [review]
